nextcord/cooldowns/cooldown.py

```diff
@@ -0,0 +1,253 @@
+        limit: int
+            How many call's can be made in the time
+            period specified by ``per``
+        per: float
+            The time period related to ``limit``
+        bucket: Optional[Bucket]
+            The :class:`Bucket` implementation to use
+            as a bucket to separate cooldown buckets.
+
+            Defaults to :class:`CooldownBucket.all`
+        func: Optional[Callable]
+            The function this cooldown is attached to
+        """
+        bucket = bucket or CooldownBucket.all
+        # See TimesPer for inspo, however it doesnt
+        # use that due to the required changes and flexibility
+        self.limit: int = limit
+        self.per: float = per
+
+        self._func: Optional[Callable] = func
+        self._bucket: Bucket = bucket
+        self.loop: AbstractEventLoop = get_event_loop()
+        self.pending_reset: bool = False
+        self.last_reset_at: Optional[float] = None
+        self._last_bucket: Optional[_HashableArguments] = None
+
+        self._cache: dict[_HashableArguments, _CooldownTimesPer] = {}
+
+    async def __aenter__(self) -> "Cooldown":
+        bucket: _CooldownTimesPer = self._get_cooldown_for_bucket(self._last_bucket)
+        async with bucket:
+            return self
+
+    async def __aexit__(self, *_) -> None:
+        ...
+
+    def __call__(self, *args, **kwargs):
+        self._last_bucket = self.get_bucket(*args, **kwargs)
+        return self
+
+    def _get_cooldown_for_bucket(self, bucket: _HashableArguments):
+        try:
+            return self._cache[bucket]
+        except KeyError:
+            _bucket = _CooldownTimesPer(self.limit, self.per, self)
+            self._cache[bucket] = _bucket
+            return _bucket
+
+    def get_bucket(self, *args, **kwargs) -> _HashableArguments:
+        """
+        Return the given bucket for some given arguments.
+
+        This uses the underlying :class:`CooldownBucket`
+        and will return a :class:`_HashableArguments`
+        instance which is inline with how Cooldown's function.
+
+        Parameters
+        ----------
+        args: Any
+            The arguments to get a bucket for
+        kwargs: Any
+            The keyword arguments to get a bucket for
+
+        Returns
+        -------
+        _HashableArguments
+            An internally correct representation
+            of a bucket for the given arguments.
+
+            This can then be used in :meth:`Cooldown.reset` calls.
+        """
+        data = self._bucket.process(*args, **kwargs)
+        if self._bucket is CooldownBucket.all:
+            return _HashableArguments(*data[0], **data[1])
+
+        elif self._bucket is CooldownBucket.args:
+            return _HashableArguments(*data)
+
+        elif self._bucket is CooldownBucket.kwargs:
+            return _HashableArguments(**data)
+
+        return _HashableArguments(data)
+
+    def reset(self, bucket: Optional[_HashableArguments] = None) -> None:
```

Comment:
This should ask for bucketprotocol?

Reply:
No? As a cooldown can only have one bucket protocol, and thats ``self.__bucket``

This resets the cooldown on a specfic Cooldown for x args